packages/@o3r/rules-engine/src/engine/ruleset-executor.ts

```diff
@@ -195,38 +195,58 @@ export class RulesetExecutor {
     throw new Error(`Unknown condition block met : ${JSON.stringify(nestedCondition)}`);
   }
 
+  // TODO: make private and move to top of the class in a separate PR to make the diff more meaningful in this one
   /**
    * Plug ruleset to fact streams and trigger a first evaluation
-   *
-   * @param ruleset
    */
   public prepareRuleset() {
-    const rulesWithContext: Rule[] = [];
-    const rulesWithoutContext: Rule[] = [];
+    const inputFactsForRule: Record<string, string[]> = {};
+    const exploreObject = (currentObj: any, ruleInputFacts: Set<string>) => {
```

Comment:
I am not sure to understand the purpose of this part :S.
Why do you need it as you already have the list of facts available for the ruleset?

Reply:
The goal is to compute the list of input facts and no longer take into account the value of the `inputFacts` coming from the CMS.
We have even removed them from the showcase app ruleset https://github.com/AmadeusITGroup/otter/pull/835/files#diff-1c83c4284e0d72e53c44790c5bdd50daedb84c30b8e703c4e0552ecf2ab870ae

Reply:
Concerning the types for the input, we support nested rules and the method will iterate on many types of Objects or Arrays.

Here's an example of a nested rule:
```typescript
        {
          'id': '6e8t54h6s4e-6erth46sre8th4-d46t8s13t5j2',
          'name': 'the second rule',
          'inputRuntimeFacts': ['UI_FACT_2'],
          'inputFacts': ['cart'],
          'outputRuntimeFacts': [],
          'rootElement': {
            'elementType': 'RULE_BLOCK',
            'blockType': 'IF_ELSE',
            'condition': {
              'any': [
                {
                  'lhs': {
                    'type': 'RUNTIME_FACT',
                    'value': 'UI_FACT_2'
                  },
                  'operator': 'equals',
                  'rhs': {
                    'type': 'LITERAL',
                    'value': false
                  }
                }
              ]
            },
            'successElements': [
              {
                'elementType': 'ACTION',
                'actionType': 'UPDATE_LOCALISATION',
                'key': 'my.loc.key2.success',
                'value': 'my.loc.value2.success'
              },
              {
                'elementType': 'ACTION',
                'actionType': 'UPDATE_LOCALISATION',
                'key': 'my.loc.key3.success',
                'value': 'my.loc.value3.success'
              },
              {
                'elementType': 'RULE_BLOCK',
                'blockType': 'IF_ELSE',
                'condition': {
                  'any': [
                    {
                      'lhs': {
                        'type': 'FACT',
                        'value': 'cart',
                        'path': '$.xmasHampers[0].hamperItems[1].id'
                      },
                      'operator': 'equals',
                      'rhs': {
                        'type': 'LITERAL',
                        'value': 'foieGras'
                      }
                    }
                  ]
                },
                'successElements': [
                  {
                    'elementType': 'ACTION',
                    'actionType': 'UPDATE_LOCALISATION',
                    'key': 'my.loc.key4.success',
                    'value': 'my.loc.value4.success'
                  }
                ],
                'failureElements': [
                  {
                    'elementType': 'ACTION',
                    'actionType': 'UPDATE_LOCALISATION',
                    'key': 'my.loc.key5.failure',
                    'value': 'my.loc.value5.failure'
                  }
                ]
              },
              {
                'elementType': 'ACTION',
                'actionType': 'UPDATE_LOCALISATION',
                'key': 'my.loc.key6.success',
                'value': 'my.loc.value6.success'
              }
            ],
            'failureElements': [
              {
                'elementType': 'ACTION',
                'actionType': 'SET_FACT',
                'fact': 'UI_FACT_2',
                'value': false
              }
            ]
          }
        }
```

The method will find the `cart` fact in a `successElements` array.